src/core/network/client.ts

```diff
@@ -31,8 +31,10 @@ const WEBHOOK_BLACKLIST = [
 
 // eslint-disable-next-line @typescript-eslint/no-namespace
 namespace ApiClient {
+  export type Agent = RequestInit['agent']
   export interface Options {
-    agent?: https.Agent | http.Agent
+    agent?: Agent
+    attachmentAgent?: Agent
```

Comment:
On second thought, we could force people to confront the issue head-on, as @septs suggested in https://github.com/telegraf/telegraf/pull/1216/files#diff-730d79a9ad7b5cb013c7df3ec05108beb986853d97a329c213700e8f2aab83c6R40:
```suggestion
    attachmentAgent?: (parsedURL: URL) => http.Agent
```

Reply:
But then, some people might naively create new agent per request, which defeats the point...

Reply:
I don't like using a factory here, that's not too elegant.

Currently, it is possible to set the correct agent by assigning it to `bot.telegram.options.attachmentAgent` before making a corresponding call. One way would be to document this explicitly in the docs, and point users to the issue of different protocols when introducing it.

Another way would be to add the good old router, but that's probably overdoing it and almost as bad as regenerating the agent per request:
```ts
<A extends Record<string, Agent>>setAttachmentAgents(routeFn: (parsed: URL) => keyof A, agents: A): void
```

We could also provide two options, one for HTTP and one for HTTPS, but that's also really weird, because there's some many more protocol that we would then somehow ignore.

I think I actually favor a good explanation in the docs here. If someone messes around with proxies, they have to know what they're doing. I'm

Reply:
> Currently, it is possible to set the correct agent by assigning it to `bot.telegram.options.attachmentAgent`

That's bad. Instead, we could introduce `attachmentAgent` option to `callApi` next to `signal`... but this still wouldn't solve the mixed protocol media group issue...

Let's just wait until it becomes an issue and someone more experienced offers a good solution.

Reply:
what is good solution?

Reply:
@wojpawlik if you not accept, you should provide "good solution"

You reject it without solving it

Reply:
@wojpawlik how about we remove `attachmentAgent` from the options one may supply, but leave it supported in `ApiClient`? We leave it undocumented, but users who have to have a proxy can still rely on `telegraf`.

I would be uncomfortable with not providing any solution whatsoever for this in 4.0, hence effectively excluding every user from sending files that has to use a proxy for some reason.

Reply:
> excluding every user from sending files that has to use a proxy for some reason

They're free to suggest a solution or continue using 3.38.0.

Reply:
So then what's the answer to the suggestion I just gave you?

Reply:
The answer is, types are documentation, so I fail to see how you could leave it in, really undocumented.

Reply:
By only allowing a direct assignment to `bot.telegram.options.attachmentAgent` but not supplying this option during setup. This is not 100 % undocumented ofc, but I assume that no one really modifies the internal options at runtime anyway.

And yes, if someone sets arbitrary values without outside of the public API that are four levels deep into the library, then I find it OK that this can break things :)

Reply:
\* `ctx.telegram.options.attachmentAgent`

`Telegram::options` is `public`, so it'd still be documented and impossible to remove before 5.0.

Reply:
Yeah but that's exactly what I'm suggesting … we can literally just `Omit` that one property in the place where the options can be passed. Then no one can do it, only afterwards by explicit assignments (cause that's where we don't `Omit` it).

I don't really see where you're going tbh. We either have to explain to people that modify this option accidentally (if that's even possible) why their bot broke, or we have to reintroduce this feature at a later point anyway and then ship a bugfix release so 4.0 is usable behind the GFW.

If all you're concerned about is just dealing with the issues opened by devs who accidentally write
`ctx.telegram.options.attachmentAgent = new https.Agent({ keepAlive: true, keepAliveMsecs: 10000 })` and don't know what this does, then I can offer to take care of them.

Reply:
try fetching check error name is `ERR_INVALID_PROTOCOL`, wrap a new error (provide an faq documentation link)
`attachmentAgent` need is factory function. (may be)
https://github.com/telegraf/telegraf/blob/28a893e6e4157731b2e2081fca191790090e3ea4/src/core/network/client.ts#L213-L221

Reply:
@KnorpelSenf I'm concerned about inventing a better solution later and still having to support the poor workaround.

> we have to reintroduce this feature at a later point anyway and then ship a bugfix release so 4.0 is usable behind the GFW

Yes, I'm fine with this.

Reply:
Dealing with errors is messy (= not well-typed) @septs, I'd rather not.